File: src/ruby/gnome_sort.rb
# frozen_string_literal: true

# Algoritmo Gnome sort em Ruby
# Carlos Alves
# https://github.com/EuCarlos

def gnome_sort(array)
  count = 0

  while count < array.length - 1
    if array[count] > array[count + 1]
      array[count + 1], array[count] = array[count], array[count + 1]
      count -= 2 if count.positive?
    end
    count += 1
  end

  array
end

def main
  array = [10, 8, 4, 3, 1, 9, 0, 2, 7, 5, 6]
  print(gnome_sort(array))
end

main

File: src/java/GnomeSort.java
import java.util.Arrays;

public class GnomeSort {
  public static void main(String[] args) {
    int vetor[] = {10, -10, -9, 9, 8, -8, 7, -7, 6, -6, 5, -5, 4, -4, 3, -3, 2, -2, 1, -1, 0};
    System.out.println("Vetor antes da ordenação:" + Arrays.toString(vetor));

    gnomeSort(vetor);
    System.out.println(
        "Vetor após ordenação com o algoritmo Gnome Sort: " + Arrays.toString(vetor));
  }

  static void gnomeSort(int[] vetor) {
    int i = 0;
    while (i < vetor.length) {
      if (i == 0 || vetor[i] >= vetor[i - 1]) {
        i++;

      } else {
        int auxiliar = vetor[i];
        vetor[i] = vetor[i - 1];
        vetor[i - 1] = auxiliar;

        i--;
      }
    }
  }
}
